add best practices on how to get near real-time data with dbt (#8478)
fishtown-analytics/docs.getdbt.com: website/docs/best-practices/how-we-handle-real-time-data/1-intro.md

---
title: "Near real-time data in dbt"
id: "1-intro"
description: Learn how to achieve near real-time data freshness with dbt through streaming ingestion and frequent transformations
hoverSnippet: Learn how to achieve near real-time data freshness with dbt
---

By design, dbt is batch-oriented with jobs having a defined start and end time. But did you know that you can also use dbt to get near real-time data by combining your data warehouse's continuous ingestion with frequent dbt transformations?

> No worries, this guide covers multiple patterns for achieving near real-time data freshness with dbt:

Reviewer: I guess this is more a question we should address as a team when planning for the future, but do we intend to maintain the informal tone often found in the Best Practices (which are sometimes written more like blogs), or should we style them similar to /docs? 

At the very least I think we should avoid idiomatic language

```suggestion
This guide covers multiple patterns for achieving near real-time data freshness with dbt:
```

Author: i like the tone informal imho - it's best practices and opinionated rather than formal. it's very 'dbt' in the sense that its purpose is to share best practices, while also maintaining the curated, dbt flair that we all have seen over time. i see it as an extension of dbt summit vibes where we're sharing knowledge on how to best use dbt based on what's worked for us or others.  we have over 120 instances of 'informal' language across all the docs as well.

but we can remove the no worries - this is def a good topic if you want to add to the agenda

> 1. [Incremental patterns](/best-practices/how-we-handle-real-time-data/2-incremental-patterns) &mdash; `merge` strategies, CDC, and microbatch processing

Reviewer 2: ```suggestion
1. [Incremental patterns](/best-practices/how-we-handle-real-time-data/2-incremental-patterns) &mdash; `merge` strategies, Change Data Capture (CDC), and microbatch processing
```

> 1. [Incremental patterns](/best-practices/how-we-handle-real-time-data/2-incremental-patterns) - `merge` strategies, CDC, and microbatch processing

Reviewer: The - in this section should be emdashes for style
```suggestion
1. [Incremental patterns](/best-practices/how-we-handle-real-time-data/2-incremental-patterns) &mdash; `merge` strategies, CDC, and microbatch processing
```

> 2. [Warehouse-native features](/best-practices/how-we-handle-real-time-data/3-warehouse-native-features) - When to use dynamic tables and materialized views

Author: ```suggestion
2. [Warehouse-native features](/best-practices/how-we-handle-real-time-data/3-warehouse-native-features) &mdash; When to use dynamic tables and materialized views
```

> 3. [Lambda views pattern](/best-practices/how-we-handle-real-time-data/4-lambda-views) - Combining batch and real-time data in a single view

Author: ```suggestion
3. [Lambda views pattern](/best-practices/how-we-handle-real-time-data/4-lambda-views) &mdash; Combining batch and real-time data in a single view
```

> 4. [Views-only pattern](/best-practices/how-we-handle-real-time-data/5-views-only-pattern) - Maximum freshness for lightweight transformations

Author: ```suggestion
3. [Lambda views pattern](/best-practices/how-we-handle-real-time-data/4-lambda-views) &mdash; Combining batch and real-time data in a single view
4. [Views-only pattern](/best-practices/how-we-handle-real-time-data/5-views-only-pattern) &mdash; Maximum freshness for lightweight transformations
```

> 5. [Operational considerations](/best-practices/how-we-handle-real-time-data/6-operational-considerations) - Challenges, risks, and cost management

Author: ```suggestion
5. [Operational considerations](/best-practices/how-we-handle-real-time-data/6-operational-considerations) &mdash; Challenges, risks, and cost management
```

Each pattern includes practical code examples, use cases, and tradeoffs to help you choose the right approach.

> Anyone can check out this guide, but it's primarily for data engineers and architects who are looking to achieve near real-time data freshness with dbt.

Reviewer: ```suggestion
Anyone can use this guide, but it's primarily for data engineers and architects who want to achieve near real-time data freshness with dbt.
```

## Where does dbt fit?

There are two main ways to use dbt to get near real-time data:

> - For near real-time (1-15 minutes) &mdash; dbt excels at this and is well-suited for most operational dashboards.

Reviewer 3: To be even more cautious i'm going to change this to (5 - 15 minutes)

> - For near real-time (1-15 minutes) &mdash; dbt excels at this and is well-suited for most operational dashboards.

Author: ```suggestion
- For near real-time (5 - 15 minutes) &mdash; dbt excels at this and is well-suited for most operational dashboards.
```

- For true real-time (sub-second) &mdash; This requires dedicated streaming databases (ClickHouse, Materialize, Rockset, and so on) in front of or alongside dbt; dbt still owns “analytic” tables and history but not the ultra‑low‑latency read path.

## How dbt achieves near real-time data
To achieve real-time data with dbt, we recommend using a two-layer architecture:

#### Ingestion layer

Continuous data landing using your data warehouse's streaming ingestion features. 
   

> Streaming ingestion features such as [streaming tables](https://docs.databricks.com/en/sql/load-data-streaming-table.html), [Snowpipe](https://docs.snowflake.com/en/user-guide/snowpipe-streaming/data-load-snowpipe-streaming-overview), or [Storage Write API](https://docs.cloud.google.com/bigquery/docs/write-api-streaming) work well for this. To find streaming ingestion features for your warehouse, refer to the [additional resources](#additional-resources) section.

Reviewer 2: additional resources link not working

> Streaming ingestion features like [streaming tables](https://docs.databricks.com/en/sql/load-data-streaming-table.html), [Snowpipe](https://docs.snowflake.com/en/user-guide/snowpipe-streaming/data-load-snowpipe-streaming-overview), or [Storage Write API](https://docs.cloud.google.com/bigquery/docs/write-api-streaming) are a great way to do this. To check streaming ingestion features for your warehouse, check the [additional resources](#additional-resources) section.

Reviewer: ```suggestion
Streaming ingestion features such as [streaming tables](https://docs.databricks.com/en/sql/load-data-streaming-table.html), [Snowpipe](https://docs.snowflake.com/en/user-guide/snowpipe-streaming/data-load-snowpipe-streaming-overview), or [Storage Write API](https://docs.cloud.google.com/bigquery/docs/write-api-streaming) work well for this. To find streaming ingestion features for your warehouse, refer to the [additional resources](#additional-resources) section.
```

#### dbt transformation layer

> Run dbt transformations every few minutes to transform that data, and you can use materialized views or dynamic tables for the lowest-latency reporting.

Reviewer: It feels like we're saying the same thing twice in rapid succession, but I'm struggling to come up with another term. Maybe `process`? Does that look ok?
```suggestion
Run dbt every few minutes to transform the data, and use materialized views or dynamic tables for the lowest-latency reporting.
```

Author: good shout. going to drop transformations and keep transform.

Specific transformation approaches include:
- [Incremental models](/docs/build/incremental-models-overview) with merge or append strategies
- [Microbatch incremental strategy](/docs/build/incremental-microbatch) for large time-series tables
- Jobs scheduled very frequently (like every 5 minutes)

> - Dynamic tables](/reference/resource-configs/snowflake-configs#dynamic-tables) or [materialized views](/reference/resource-configs/snowflake-configs#materialized-views) with short refresh intervals

Reviewer 2: ```suggestion
- [Dynamic tables](/reference/resource-configs/snowflake-configs#dynamic-tables) or [materialized views](/reference/resource-configs/snowflake-configs#materialized-views) with short refresh intervals
```

Author: oh good catch!

> - Use [dynamic tables](/reference/resource-configs/snowflake-configs#dynamic-tables) or [materialized views](/reference/resource-configs/snowflake-configs#materialized-views) with short refresh intervals

Reviewer: ```suggestion
- Dynamic tables](/reference/resource-configs/snowflake-configs#dynamic-tables) or [materialized views](/reference/resource-configs/snowflake-configs#materialized-views) with short refresh intervals
```

## Key recommendations

> Here are some key recommendations to help you achieve near real-time data freshness with dbt:

Reviewer: ```suggestion
The following are some key recommendations to help you achieve near real-time data freshness with dbt:
```

> - Ingest data continuously &mdash; Use your warehouse's native streaming or micro-batch ingestion to land raw data as soon as it arrives.

Reviewer 2: Can we replace the emdash with a `:` to introduce the explanation? This feels more scannable than an emdash.

> - Transform with dbt on a frequent schedule &mdash; Schedule dbt jobs to run as often as your business needs allow (e.g., every 1–15 minutes). Balance freshness with cost and resource constraints.

Reviewer: ```suggestion
- Transform with dbt on a frequent schedule &mdash; Schedule dbt jobs to run as often as your business needs allow (for example, every 1–15 minutes). Balance freshness with cost and resource constraints.
```

> - Materialized views & dynamic tables &mdash; For the lowest-latency reporting, use materialized views or dynamic tables. These can be refreshed as frequently as every minute.

Reviewer: We're switching between & and "and" in this section
```suggestion
- Materialized views and dynamic tables &mdash; For the lowest-latency reporting, use materialized views or dynamic tables. These can be refreshed as frequently as every minute.
```

> - Incremental models & microbatching &mdash; Use dbt's incremental models to process only new or changed data, keeping transformations efficient and scalable.

Reviewer: ```suggestion
- Incremental models and microbatching &mdash; Use dbt's incremental models to process only new or changed data, keeping transformations efficient and scalable.
```
